src/offspot_demo/deploy.py

```diff
@@ -1,2 +1,345 @@
+        size = resp.headers.get("Content-Length")
+        etag = resp.headers.get("ETag", "").replace('"', "").replace("'", "")
+    if not size or not etag:
+        return S3CompatibleETag("", 0, 0, 0)
+
+    size = int(size)
+
+    # single part etag
+    if "-" not in etag:
+        return S3CompatibleETag(etag, 1, size, size)
+
+    digest, nb_parts = etag.split("-", 1)
+    nb_parts = int(nb_parts)
+
+    size_in_mib = size // ONE_MIB
+    if size_in_mib % nb_parts != 0:
+        parts_size = size // (nb_parts - 1)
+    else:
+        parts_size = size // nb_parts
+    # round to MiB
+    parts_size = parts_size // ONE_MIB * ONE_MIB
+
+    return S3CompatibleETag(digest, nb_parts, parts_size, size)
+
+
+def compute_s3etag_for(fpath: Path, digest: S3CompatibleETag):
+    """Compute-back an S3 multipart ETag using local file and info from orig ETag"""
+    concat_sum = b""
+    with open(fpath, "rb") as fh:
+        for _ in range(digest.nb_parts):
+            sum_ = hashlib.md5(fh.read(digest.parts_size), usedforsecurity=False)
+            concat_sum += sum_.digest()
+    concat_hex = hashlib.md5(concat_sum, usedforsecurity=False).hexdigest()
+    return f"{concat_hex}-{digest.nb_parts}"
+
+
+def download_file_into(url: str, dest: Path, digest: S3CompatibleETag) -> int:
+    """Download url into dest using aria2c, validating checksum"""
+
+    dest.parent.mkdir(parents=True, exist_ok=True)
+
+    # download into a temp folder next to target
+    with TemporaryDirectory(
+        suffix=".aria2", dir=dest.parent, ignore_cleanup_errors=True, delete=True
+    ) as tmpdir:
+        tmp_dest = Path(tmpdir).joinpath("image.img")
+
+        args = ["aria2c", "--dir", str(tmp_dest.parent), "--out", "image.img"]
+        # single part checksum, let aria2 handle checksum validation
+        if digest.is_singlepart:
+            args += ["--checksum", digest.checksum]
+        args += [url]
+        aria2 = subprocess.run(args, check=False)
+
+        if aria2.returncode != 0:
+            logger.error("Failed to download with aria2c: {aria2.returncode}")
+            return aria2.returncode
+
+        if digest.is_multipart:
+            logger.info(">> verify checksum…")
+
+            computed = compute_s3etag_for(fpath=tmp_dest, digest=digest)
+            if computed != digest.etag:
+                logger.error(
+                    f"MD5 checksum validation failed: {computed=} != {digest.etag}"
+                )
+                return 32
+        # move to destination (should be safe as we're in sub of parent)
+        tmp_dest.rename(dest)
+
+    logger.info("Download completed")
+    return 0
+
+
+def deploy_url(url: str, *, reuse_image: bool):
+    """Deploy from an URL
+
+    Parameters:
+        url: the image URL (ideally hosted on S3)
+        reuse_image: whether to not remove existing image file and skip download (dev)
+    """
+    rc = do_deploy_url(url, reuse_image=reuse_image)
+    if rc:
```

Comment:
Shouldn't we use a context manager to be sure to do the cleanup, no matter which exception might occur? (or at least try...except?)

Reply:
Indeed !